CRM/Contribute/Form/ContributionBase.php

```diff
@@ -1234,30 +1235,16 @@ protected function buildMembershipBlock(
               $memType['id'], $javascriptMethod
             );
             if ($cid) {
-              $membership = new CRM_Member_DAO_Membership();
-              $membership->contact_id = $cid;
-              $membership->membership_type_id = $memType['id'];
-
-              //show current membership, skip pending and cancelled membership records,
-              //because we take first membership record id for renewal
-              $membership->whereAdd('status_id != 5 AND status_id !=6');
```

Comment:
In getAllContactMembership it excludes 'pending' but not 'cancelled' from what I can see so this seems like it would lose that

Reply:
You're right.  I was only thinking in terms of the Pending-then-Cancelled edge case, and not Cancelled generally.  I'll update this PR.

Reply:
OK ping me when you have - I think this is otherwise mergeable

Reply:
Hey @eileenmcnaughton I took a closer look at this and I think I want to stand by the code as written.

We're identifying lifetime memberships because they're the only memberships that have a start date but no end date.  That's what the test I wrote is checking for.

The true edge case here is *cancelled lifetime* membership.  My code flags that as a lifetime membership - your suggestion would consider it to NOT be a lifetime membership.

I can be convinced by others to change my definition, but my inclination is that my definition is a good one.